### ISSUE TITLE
Exception throwing during rescan

### ISSUE THREAD

**Issue by Christewart** [bitcoin-s/bitcoin-s#3869]
When rescanning my wallet on d393848cc204b92edc3c200faab05c1f72bf3e75 with my wallet connected via neutrino node to the bitcoin network. It appears my rescan died on block height `685523` due to the dlc module of the wallet

```
2021-12-02T19:12:33UTC INFO [DLCWallet$DLCWalletImpl] Adding UTXO to wallet: c1e817693548348bdc11ec6ef3a21a34039c14e89209ee94a034641cbfd92d13:2 amt=9822359 sats
2021-12-02T19:12:33UTC INFO [DLCWallet$DLCWalletImpl] Successfully inserted UTXO c1e817693548348bdc11ec6ef3a21a34039c14e89209ee94a034641cbfd92d13:2 into DB
2021-12-02T19:12:33UTC INFO [DLCWallet$DLCWalletImpl] Processing tx 132dd9bf1c6434a094ee0992e8149c03341aa2f36eec11dc8b3448356917e8c1 for 1 DLC(s)
2021-12-02T19:12:33UTC INFO [DLCWallet$DLCWalletImpl] 1 txos are now confirmed!
2021-12-02T19:12:40UTC INFO [DLCWallet$DLCWalletImpl] Finished processing of block=00000000000000000001e6900be9b4987ba898758cf8c58f4626e3528bc6947c. It took 12190ms
2021-12-02T19:12:40UTC INFO [DLCWallet$DLCWalletImpl] Found 1 matches from start=685323 to end=685422
2021-12-02T19:12:40UTC INFO [DLCWallet$DLCWalletImpl] Requesting 0 block(s)
2021-12-02T19:12:40UTC INFO [DLCWallet$DLCWalletImpl] Found 0 matches from start=685423 to end=685522
2021-12-02T19:12:40UTC INFO [DLCWallet$DLCWalletImpl] Found a match in block 685523
2021-12-02T19:12:41UTC INFO [DLCWallet$DLCWalletImpl] Requesting 1 block(s)
2021-12-02T19:12:44UTC INFO [DataMessageHandler] Received block message with hash 000000000000000000085a8452a22250569563f4adbd23cb3c22b116ee934296
2021-12-02T19:12:44UTC INFO [DLCWallet$DLCWalletImpl] Processing block=000000000000000000085a8452a22250569563f4adbd23cb3c22b116ee934296
2021-12-02T19:12:48UTC INFO [DLCWallet$DLCWalletImpl] Adding UTXO to wallet: 0aa02c1c143435c0db858b791d64a1ccb2608d24835aed34038c737857010d82:0 amt=100000 sats
2021-12-02T19:12:48UTC INFO [DLCWallet$DLCWalletImpl] Successfully inserted UTXO 0aa02c1c143435c0db858b791d64a1ccb2608d24835aed34038c737857010d82:0 into DB
2021-12-02T19:12:48UTC INFO [DLCWallet$DLCWalletImpl] Processing tx 820d015778738c0334ed5a83248d60b2cca1641d798b85dbc03534141c2ca00a for 1 DLC(s)
2021-12-02T19:12:49UTC ERROR [DLCWallet$DLCWalletImpl] Error processing of block=000000000000000000085a8452a22250569563f4adbd23cb3c22b116ee934296.
java.util.NoSuchElementException: None.get
	at scala.None$.get(Option.scala:627)
	at scala.None$.get(Option.scala:626)
	at org.bitcoins.dlc.wallet.internal.DLCTransactionProcessing.$anonfun$calculateAndSetOutcome$13(DLCTransactionProcessing.scala:150)
	at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:431)
	at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:63)
	at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:100)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:94)
	at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:100)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:49)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:48)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
2021-12-02T19:12:49UTC ERROR [DataMessageHandler] onBlockReceived Callback failed with error: 
java.util.NoSuchElementException: None.get
	at scala.None$.get(Option.scala:627)
	at scala.None$.get(Option.scala:626)
	at org.bitcoins.dlc.wallet.internal.DLCTransactionProcessing.$anonfun$calculateAndSetOutcome$13(DLCTransactionProcessing.scala:150)
	at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:431)
	at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:63)
	at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:100)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:94)
	at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:100)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:49)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:48)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
2021-12-02T19:12:49UTC INFO [DLCWallet$DLCWalletImpl] Finished rescanning the wallet. It took 2006532ms
2021-12-02T19:13:10UTC INFO [DataMessageHandler] Received headers message with 1 headers
2021-12-02T19:13:11UTC INFO [DataMessageHandler] Starting to fetch filter headers in data message handler
2021-12-02T19:13:11UTC INFO [DataMessageHandler] Got 1 compact filter header hashes
2021-12-02T19:13:11UTC INFO [DataMessageHandler] Done syncing filter headers, beginning to sync filters in datamessagehandler
2021-12-02T19:13:11UTC INFO [PeerMessageSender] Requesting compact filters from FilterSyncMarker(startHeight=712275, stopBlockHash=00000000000000000007dfcd40f1efea4e0a81894189c90d78364b271f075332)
2021-12-02T19:13:12UTC INFO [DataMessageHandler] We are synced
2021-12-02T19:13:12UTC INFO [DataMessageHandler] Processing 1 filters
2021-12-02T19:14:33UTC INFO [DataMessageHandler] Received headers message with 1 headers
2021-12-02T19:14:33UTC INFO [DataMessageHandler] Starting to fetch filter headers in data message handler
2021-12-02T19:14:33UTC INFO [DataMessageHandler] Got 1 compact filter header hashes
2021-12-02T19:14:33UTC INFO [DataMessageHandler] Done syncing filter headers, beginning to sync filters in datamessagehandler
2021-12-02T19:14:34UTC INFO [PeerMessageSender] Requesting compact filters from FilterSyncMarker(startHeight=712276, stopBlockHash=00000000000000000002739d3caac26a350da8546e2b5ded08d9922efc065be1)
2021-12-02T19:14:34UTC INFO [DataMessageHandler] We are synced
2021-12-02T19:14:34UTC INFO [DataMessageHandler] Processing 1 filters
2021-12-02T19:17:49UTC INFO [DataMessageHandler] Received headers message with 1 headers
2021-12-02T19:17:50UTC INFO [DataMessageHandler] Starting to fetch filter headers in data message handler
2021-12-02T19:17:50UTC INFO [DataMessageHandler] Got 1 compact filter header hashes
2021-12-02T19:17:50UTC INFO [DataMessageHandler] Done syncing filter headers, beginning to sync filters in datamessagehandler
2021-12-02T19:17:50UTC INFO [PeerMessageSender] Requesting compact filters from FilterSyncMarker(startHeight=712277, stopBlockHash=000000000000000000054df5b47f09c02a2ac2397ecfd20c54ef2ebe5351c354)
2021-12-02T19:17:51UTC INFO [DataMessageHandler] We are synced
2021-12-02T19:17:51UTC INFO [DataMessageHandler] Processing 1 filters
2021-12-02T19:23:57UTC INFO [DataMessageHandler] Received headers message with 1 headers
2021-12-02T19:23:57UTC INFO [DataMessageHandler] Starting to fetch filter headers in data message handler
2021-12-02T19:23:57UTC INFO [DataMessageHandler] Got 1 compact filter header hashes
2021-12-02T19:23:57UTC INFO [DataMessageHandler] Done syncing filter headers, beginning to sync filters in datamessagehandler
2021-12-02T19:23:57UTC INFO [PeerMessageSender] Requesting compact filters from FilterSyncMarker(startHeight=712278, stopBlockHash=000000000000000000047a16c83a4dd0bbcad9e0ceff1f2ddfba4df3b46f8201)
2021-12-02T19:23:58UTC INFO [DataMessageHandler] We are synced
2021-12-02T19:23:58UTC INFO [DataMessageHandler] Processing 1 filters
2021-12-02T19:24:23UTC INFO [DataMessageHandler] Received headers message with 1 headers
2021-12-02T19:24:23UTC INFO [DataMessageHandler] Starting to fetch filter headers in data message handler
2021-12-02T19:24:23UTC INFO [DataMessageHandler] Got 1 compact filter header hashes
2021-12-02T19:24:23UTC INFO [DataMessageHandler] Done syncing filter headers, beginning to sync filters in datamessagehandler
2021-12-02T19:24:24UTC INFO [PeerMessageSender] Requesting compact filters from FilterSyncMarker(startHeight=712279, stopBlockHash=000000000000000000055e7c761f70b5fc1e7ba2b0647a40bc068d904abd5fef)
2021-12-02T19:24:24UTC INFO [DataMessageHandler] We are synced
2021-12-02T19:24:24UTC INFO [DataMessageHandler] Processing 1 filters
2021-12-02T19:36:49UTC INFO [DLCWallet$DLCWalletImpl] Creating DLC Offer
2021-12-02T19:36:50UTC INFO [DLCWallet$DLCWalletImpl] Spending UTXOs: 7f19a7018dedb563751afeaaf8b56ec64844ac8016ca5f4c66ede8c9b5fd44ab:0, 506bc8db921f563b40a01ba7c4c97533daa79cbdaca907413e17161cbe939f9d:0, d87fedabac6c804a75afcdc8fccb93315e9854a10487cdf31407adaff98b69c4:1
2021-12-02T19:36:50UTC INFO [DLCWallet$DLCWalletImpl] UTXO 0 details: TransactionOutput(100000 sats,wpkh(cd8dace1760fc074cd84627d243440742a45c7d5))
2021-12-02T19:36:50UTC INFO [DLCWallet$DLCWalletImpl] UTXO 1 details: TransactionOutput(40000 sats,wpkh(d1cc8afbb2c771b2356a53a60212d533db77a511))
2021-12-02T19:36:50UTC INFO [DLCWallet$DLCWalletImpl] UTXO 2 details: TransactionOutput(1179 sats,wpkh(52584eb56f12d421e5efc49ecddc5eb873958cc0))
2021-12-02T19:36:50UTC INFO [DLCWallet$DLCWalletImpl] Created offer with tempContractId 0721ac7b6c20e94883ade64622999c376906a166eb005d22c8f2e638088f9b48
2021-12-02T19:56:27UTC INFO [DataMessageHandler] Received headers message with 1 headers
2021-12-02T19:56:27UTC INFO [DataMessageHandler] Starting to fetch filter headers in data message handler
2021-12-02T19:56:28UTC INFO [DataMessageHandler] Got 1 compact filter header hashes
2021-12-02T19:56:28UTC INFO [DataMessageHandler] Done syncing filter headers, beginning to sync filters in datamessagehandler
2021-12-02T19:56:28UTC INFO [PeerMessageSender] Requesting compact filters from FilterSyncMarker(startHeight=712280, stopBlockHash=00000000000000000001a4d48f5843ad36df5d2e451fa08c1cb65dcec52c6f51)
2021-12-02T19:56:29UTC INFO [DataMessageHandler] We are synced
2021-12-02T19:56:29UTC INFO [DataMessageHandler] Processing 1 filters
```

**Comment by Christewart** [comment]
This is where this exception is coming from 

```
org.bitcoins.dlc.wallet.internal.DLCTransactionProcessing.$anonfun$calculateAndSetOutcome$12(DLCTransactionProcessing.scala:155)
```

https://github.com/bitcoin-s/bitcoin-s/blob/8857af2b66e8d493814d0bb7b817d8d46579709d/dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/internal/DLCTransactionProcessing.scala#L155

**Comment by benthecarman** [comment]
I believe this is happening because after a DLC has been closed we delete all of the no longer needed CET sigs. So when we rescan and it processes the closing transaction of the DLC it tries to calcuate the outcome but fails because it doesn't have any of the CET sigs.

This should be able to be fixed by just having it skip doing the processing if the state is `Claimed`, `RemoteClaimed` or `Refunded`

**Comment by Christewart** [comment]
I get this error when trying to rescan on https://github.com/bitcoin-s/bitcoin-s/pull/3965/commits/ed9adb7771f6a294b19989b3be5038012502dff7 on my desktop. It seems to be a different error that the original issue. I'm not sure if it means this issue is fixed and this is a new unique issue, or of its related to new bug in #3965 

```
2022-01-08T17:12:55UTC INFO [DLCWallet$DLCWalletImpl] Found 1 matches from start=703891 to end=703990
2022-01-08T17:12:56UTC INFO [DLCWallet$DLCWalletImpl] Requesting 0 block(s)
2022-01-08T17:12:56UTC INFO [DLCWallet$DLCWalletImpl] Found 0 matches from start=703991 to end=704090
2022-01-08T17:12:56UTC INFO [DLCWallet$DLCWalletImpl] Found a match in block 704147
2022-01-08T17:12:56UTC INFO [DLCWallet$DLCWalletImpl] Requesting 1 block(s)
2022-01-08T17:12:57UTC INFO [DataMessageHandler] Received block message with hash 0000000000000000000d495f73083e0395ecdf432d8ffebff2d06789a8489722
2022-01-08T17:12:57UTC INFO [DLCWallet$DLCWalletImpl] Processing block=0000000000000000000d495f73083e0395ecdf432d8ffebff2d06789a8489722
2022-01-08T17:13:02UTC INFO [DLCWallet$DLCWalletImpl] Adding UTXO to wallet: de13c464b401ef130c7c67b2fa456b127f4461f0978957d63a12959c3cb6e98d:1 amt=33580 sats
2022-01-08T17:13:02UTC INFO [DLCWallet$DLCWalletImpl] Successfully inserted UTXO de13c464b401ef130c7c67b2fa456b127f4461f0978957d63a12959c3cb6e98d:1 into DB
2022-01-08T17:13:02UTC INFO [DLCWallet$DLCWalletImpl] Finished processing 1 received outputs, it took=187ms
2022-01-08T17:13:03UTC ERROR [DLCWallet$DLCWalletImpl] Error processing of block=0000000000000000000d495f73083e0395ecdf432d8ffebff2d06789a8489722.
java.lang.IllegalArgumentException: requirement failed: Offer and accept (without sigs) must refer to same event, offer.tempContractId=3924d10fa61467ffb2aabc9b9d718bb8efd04a3e93f338b8eacd8bc2877252a6 tempContractId=ca253ce6d58a0a3574ac957fe3c85b664166165c06f5ed869c322baf3344056d
        at scala.Predef$.require(Predef.scala:337)
        at org.bitcoins.core.protocol.dlc.build.DLCTxBuilder.<init>(DLCTxBuilder.scala:97)
        at org.bitcoins.dlc.wallet.internal.DLCDataManagement.$anonfun$builderFromDbData$8(DLCDataManagement.scala:434)
        at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:431)
        at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:63)
        at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:100)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
        at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:94)
        at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:100)
        at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:49)
        at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:48)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
2022-01-08T17:13:03UTC ERROR [DataMessageHandler] onBlockReceived Callback failed with error:
java.lang.IllegalArgumentException: requirement failed: Offer and accept (without sigs) must refer to same event, offer.tempContractId=3924d10fa61467ffb2aabc9b9d718bb8efd04a3e93f338b8eacd8bc2877252a6 tempContractId=ca253ce6d58a0a3574ac957fe3c85b664166165c06f5ed869c322baf3344056d
        at scala.Predef$.require(Predef.scala:337)
        at org.bitcoins.core.protocol.dlc.build.DLCTxBuilder.<init>(DLCTxBuilder.scala:97)
        at org.bitcoins.dlc.wallet.internal.DLCDataManagement.$anonfun$builderFromDbData$8(DLCDataManagement.scala:434)
        at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:431)
        at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:63)
        at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:100)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
        at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:94)
        at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:100)
        at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:49)
        at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:48)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
```

**Comment by Christewart** [comment]
I suspect this issue is fixed, and there is a new issue that was found after this issue was fixed (#3969). I'm going to leave this open for now, but if we don't see this issue again we should close this issue.

**Comment by Christewart** [comment]
This is fixed on my wallet. I created a [new branch](https://github.com/Christewart/bitcoin-s-core/commits/2021-01-09-rescan-test) that removed the new serialization introduced in #3854 and the rescan worked! :tada:
